### PR TITLE
gh-107080: Fix Py_TRACE_REFS Crashes Under Isolated Subinterpreters

### PR DESCRIPTION
**Pull request by ericsnowcurrently**
The linked list of objects was a global variable, which broke isolation between interpreters, causing crashes.  To solve this, we've moved the linked list to each interpreter.

* Issue: gh-107080

### CODE REVIEW

File: Objects/object.c [hunk 1]
```diff
@@ -159,11 +159,8 @@ _PyDebug_PrintTotalRefs(void) {
    Do not call them otherwise, they do not initialize the object! */
 
 #ifdef Py_TRACE_REFS
-/* Head of circular doubly-linked list of all objects.  These are linked
```
Comment (Yhg1s) [inline Objects/object.c:162]: This comment should probably be preserved (and updated), not just removed.

File: Python/pylifecycle.c [hunk 2]
```diff
@@ -1921,12 +1921,15 @@ Py_FinalizeEx(void)
...
+        _Py_PrintReferences(tstate->interp, stderr);
     }
 
     if (dump_refs_fp != NULL) {
-        _Py_PrintReferences(dump_refs_fp);
+        _Py_PrintReferences(tstate->interp, dump_refs_fp);
     }
+
+    PyObject refchain = { 0 };
+    _Py_StealRefchain(tstate->interp, &refchain);
```
Comment (Yhg1s) [inline Python/pylifecycle.c]: Why is this necessary?

File: Objects/object.c (hunk 1 above)
Comment (ericsnowcurrently) [reply Objects/object.c:162]: I moved the comment to the header file.

File: Python/pylifecycle.c (hunk 2 above)
Comment (ericsnowcurrently) [reply Python/pylifecycle.c]: The interpreter will be deallocated before we later call `_Py_PrintReferenceAddresses()` to walk the linked list.

That said, I think there's a related problem that I still need to address.  Namely, every remaining object in the linked list will be (effectively) freed when the interpreter's allocator is finalized, so the `_ob_next` and `_ob_prev` accesses might segfault.  It likely isn't a big problem *during* runtime finalization (except perhaps for embedders) but I'll fix that regardless.

Comment (ericsnowcurrently) [reply Python/pylifecycle.c]: Hmm, I'm pretty sure that's not actually a problem.  We only call `_Py_PrintReferenceAddresses()` in `Py_FinalizeEx()` and only for the main interpreter, which is statically allocated.  Objects in the linked list might still cause a segfault but only if they would have before this change, so I'm not going to worry about that here.  I will leave a note in the code though.

**Review by Yhg1s — APPROVED** [review]
